src/category_theory/preadditive/Mat.lean

```diff
@@ -0,0 +1,458 @@
+/-
+Copyright (c) 2021 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Morrison
+-/
+import category_theory.preadditive.default
+import category_theory.preadditive.single_obj
+import category_theory.preadditive.additive_functor
+import category_theory.limits.shapes.biproducts
+import algebra.big_operators.basic
+import data.matrix.notation
+
+/-!
+# Matrices over a category.
+
+When `C` is a preadditive category, `Mat_ C` is the preadditive categoriy
+whose objects are finite tuples of objects in `C`, and
+whose morphisms are matrices of morphisms from `C`.
+
+There is a functor `Mat_.embedding : C ⥤ Mat_ C` sending morphisms to one-by-one matrices.
+
+`Mat_ C` has finite biproducts.
+
+## The additive envelope
+
+We show that this construction is the "additive envelope" of `C`,
+in the sense that any additive functor `F : C ⥤ D` to a category `D` with biproducts
+lifts to a functor `Mat_.lift F : Mat_ C ⥤ D`,
+Moreover, this functor is unique (up to natural isomorphisms) amongst functors `L : Mat_ C ⥤ D`
+such that `embedding C ⋙ L ≅ F`.
+(As we don't have 2-category theory, we can't explicitly state that `Mat_ C` is
+the initial object in the 2-category of categories under `C` which have biproducts.)
```

Comment:
This isn't the first time this pattern has showed up - I think it's maybe the third or fourth I can think of? Maybe now is a good time to think about how we could do 2-categories.

Reply:
Yes please!

Reply:
I am very keen to do this, but will make sure to point to this archived comment thread whenever someone accuses me of introducing abstract nonsense into mathlib. :-)